tcms/rpc/tests/test_environment.py

```diff
@@ -0,0 +1,95 @@
+            self.rpc_client.Environment.filter(None)
+
+
+@override_settings(LANGUAGE_CODE="en")
+class EnvironmentCreate(APITestCase):
+    def test_environment_create_with_no_perms(self):
+        self.rpc_client.Auth.logout()
+        with self.assertRaisesRegex(ProtocolError, "403 Forbidden"):
+            self.rpc_client.Environment.create({})
+
+    def test_environment_create_with_no_required_fields(self):
+        values = {}
+        with self.assertRaisesRegex(
+            XmlRPCFault, "name.*This field is required."
+        ):
+            self.rpc_client.Environment.create(values)
+
+        values["description"] = "Description"
+        with self.assertRaisesRegex(XmlRPCFault, "name.*This field is required"):
+            self.rpc_client.Environment.create(values)
+
+    def test_environment_duplicate(self):
+        values = {"name": "E7", "description": "Description"}
+        self.rpc_client.Environment.create(values)
+        with self.assertRaisesRegex(XmlRPCFault, "name.*Environment with this Name already exists"):
+            self.rpc_client.Environment.create(values)
+
+    def test_environment_create_with_chinese(self):
+        # also see https://github.com/kiwitcms/Kiwi/issues/1770
+        values = {"name": "开源中国", "description": "中国"}
+        b = self.rpc_client.Environment.create(values)
+        self.assertIsNotNone(b)
+        self.assertEqual(b["name"], "开源中国")
+        self.assertEqual(b["description"], "中国")
+
+    def test_environment_create(self):
+        values = {"name": "E8", "description": "Description."}
+        b = self.rpc_client.Environment.create(values)
+        self.assertIsNotNone(b)
+        self.assertEqual(b["name"], "E8")
+        self.assertEqual(b["description"], "Description.")
+
+
+class EnvironmentFilter(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.environment = EnvironmentFactory()
+
+    def test_environment_filter_with_non_exist_id(self):
+        self.assertEqual(0, len(self.rpc_client.Environment.filter({"pk": -9999})))
+
+    def test_environment_filter_with_id(self):
+        b = self.rpc_client.Build.filter({"pk": self.environment.pk})[0]
+        self.assertIsNotNone(b)
+        self.assertEqual(b["id"], self.environment.pk)
+        self.assertEqual(b["name"], self.environment.name)
+        self.assertEqual(b["description"], self.environment.description)
+
+    def test_build_filter_with_name(self):
+        b = self.rpc_client.Build.filter(
+            {"name": self.environment.name}
+        )[0]
+        self.assertIsNotNone(b)
+        self.assertEqual(b["id"], self.environment.pk)
+        self.assertEqual(b["name"], self.environment.name)
+        self.assertEqual(b["description"], self.environment.description)
+
+    def test_build_filter_with_name_and_description(self):
+        b = self.rpc_client.Build.filter(
+            {"name": self.environment.name, "description": self.environment.description}
+        )[0]
```

Comment:
This and all methods in this test class are copy & paste and are not actually needed. Testing is already covered inside the `TestFilterPermission` test class above. Remove `EnvironmentFilter`.

Reply:
Removed factory as well